### PR TITLE
Add meta-rpi-domd layer

### PR DESCRIPTION
**Pull request by dsemenets**
Added u-boot v2024.04
Added boot script creation procedure
Added rpi-image-minimal

### CODE REVIEW

File: meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+SUMMARY = "A small image just capable of allowing a device to boot."
+
+require recipes-core/images/core-image-minimal.bb
+
+#COMPATIBLE_MACHINE = "(falcon|condor|eagle|whitehawk|grayhawk)"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb]: please no commented code

File: meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    i2c-tools \
+    coreutils \
+    xen \
+    xen-tools \
+    xen-tools-devd \
+    xen-tools-xenstore \
+    dnsmasq \
+    openssh \
+"
+#DEPENDS += "perl-native libarchive-zip-perl-native"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb]: Why have you added libarchive-zip-perl? You don't use it. and please no commented code

File: meta-rpi-domd/recipes-core/images/image-initramfs.bb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+DESCRIPTION = "Aos initramfs"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-core/images/image-initramfs.bb]: why do we need this? Is it really aos?

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+SUMMARY = "U-boot boot script for Xen on Raspberry Pi 4"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb]: is it for raspberry 4?

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+run set_cells
+run set_addr
+fdt mknod /chosen module@0
+fdt set /chosen/module@0 compatible "xen,linux-zimage" "xen,multiboot-module"
+fdt resize
+fdt set /chosen/module@0 reg <${kernel_addr_r} 0x${filesize} >
+fdt resize
+fdt set /chosen/module@0 bootargs "reboot=w coherent_pool=1M 8250.nr_uarts=1 pci=pcie_bus_safe snd_bcm2835.enable_compat_alsa=0 snd_bcm2835.enable_hdmi=1  smsc95xx.macaddr=2C:CF:67:32:84:23 vc_mem.mem_base=0x3fc00000 vc_mem.mem_size=0x40000000  console=ttyAMA10,115200 console=tty1 root=PARTUUID=1c507a66-02 rootfstype=ext4 fsck.repair=yes rootwait quiet splash plymouth.ignore-serial-console"
+
+fdt print /chosen
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:13]: I don't think it's needed

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+print kernel_addr_r
+print filesize
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in]: same here

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# require ../../../../poky/meta/recipes-bsp/u-boot/u-boot.inc
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: please no commented code

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+# require ../../../../poky/meta/recipes-bsp/u-boot/u-boot-common.inc
+
+require recipes-bsp/u-boot/u-boot.inc
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb:1]: why create this recipe? I think u-boot_%.bbappend is enough. please remove

File: meta-rpi-domd/include/rpi-image-dev.inc [hunk 9]
```diff
@@ -0,0 +1,28 @@
+IMAGE_FEATURES += "\
+    tools-sdk \
+    tools-debug \
+    debug-tweaks \
+    dev-pkgs \
+"
+
+CORE_IMAGE_EXTRA_INSTALL += "\
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/include/rpi-image-dev.inc:8]: have you tried to build this?

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb (hunk 1 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb]: Deleted

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb (hunk 7 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: Deleted comments

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 5 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:13]: boot script reworked

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb (hunk 4 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bb]: Fixed

File: meta-rpi-domd/include/rpi-image-dev.inc (hunk 9 above)
Comment (dsemenets) [reply meta-rpi-domd/include/rpi-image-dev.inc:8]: Yes

File: meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb (hunk 2 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-core/images/rpi-image-minimal.bb]: Deleted

File: meta-rpi-domd/recipes-core/images/image-initramfs.bb (hunk 3 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-core/images/image-initramfs.bb]: Deleted recipe

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 6 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in]: boot script reworked

**Comment by dsemenets** [comment]
Fixed rpi-image-minimal
Added recipe for kernel
Fixed u-boot_2024.04.bb
Deleted image-initramfs.bb

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb (hunk 7 above)
Comment (oleksiimoisieiev) [reply meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: still see in changelog

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 5 above)
Comment (oleksiimoisieiev) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:13]: don't see changes

File: meta-rpi-domd/recipes-kernel/linux/files/0001-dts-Add-device-tree-for-xen.patch [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+diff --git a/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts b/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts
+index 1457e696f968..cbc7ee7e600e 100644
+--- a/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts
++++ b/arch/arm64/boot/dts/broadcom/bcm2712-rpi-5-b.dts
+@@ -1,2 +1,13 @@
+ // SPDX-License-Identifier: GPL-2.0
+ #include "arm/broadcom/bcm2712-rpi-5-b.dts"
++
++/ {
++	chosen {
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-kernel/linux/files/0001-dts-Add-device-tree-for-xen.patch]: chosen in domd.dts?

File: meta-rpi-domd/recipes-support/libarchive-zip-perl/libarchive-zip-perl_1.60.bb [hunk 11]
```diff
@@ -0,0 +1,20 @@
+DESCRIPTION = "Archive::Zip - Provide an interface to ZIP archive files. \
+               It provides also the crc32 binary"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-support/libarchive-zip-perl/libarchive-zip-perl_1.60.bb]: It's not used

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb (hunk 7 above)
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: so u-boot_2024.04.bb or u-boot_%.bbappend?

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 6 above)
Comment (oleksiimoisieiev) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in]: don't see any changes

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb (hunk 7 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: Sorry. Fixed

File: meta-rpi-domd/recipes-support/libarchive-zip-perl/libarchive-zip-perl_1.60.bb (hunk 11 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-support/libarchive-zip-perl/libarchive-zip-perl_1.60.bb]: Deleted

File: meta-rpi-domd/recipes-kernel/linux/files/0001-dts-Add-device-tree-for-xen.patch (hunk 10 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-kernel/linux/files/0001-dts-Add-device-tree-for-xen.patch]: What you mean?
This dts used by Xen

File: meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb (hunk 7 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/u-boot/u-boot_2024.04.bb]: Both of them.
u-boot_2024.04.bb for u-boot itself u-boot_%.bbappend has rpi5 specific overrides

**Comment by dsemenets** [comment]
Corrected u-boot
Corrected boot script

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 5 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:13]: What you mean? I still modify chosen node with address and size

**Comment by dsemenets** [comment]
Divided on several commits

File: meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in (hunk 5 above)
Comment (oleksiimoisieiev) [reply meta-rpi-domd/recipes-bsp/xt-rpi-u-boot-scr/files/boot.cmd.xen.in:13]: I mean I don't see changes in the boot script itself

File: meta-rpi-domd/conf/layer.conf [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+BBFILE_COLLECTIONS += "xt-raspberrypi-domd"
+BBFILE_PATTERN_xt-raspberrypi-domd := "^${LAYERDIR}/"
+BBFILE_PRIORITY_xt-raspberrypi-domd = "10"
+
+LAYERSERIES_COMPAT_xt-raspberrypi-domd = "kirkstone scarthgap"
+LAYERDEPENDS_xt-raspberrypi-domd = "core"
+
+BBFILES_DYNAMIC += " \
+    raspberrypi:${LAYERDIR}/dynamic-layers/raspberrypi/*/*/*.bb \
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/conf/layer.conf]: Is it needed here? I don't see any dynamic layers in meta-rpi-domd

File: meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ dtb-$(CONFIG_ARCH_BCM2835) += bcm2712-rpi-cm5-cm5io.dtb
+ dtb-$(CONFIG_ARCH_BCM2835) += bcm2712-rpi-cm5-cm4io.dtb
+diff --git a/arch/arm64/boot/dts/broadcom/bcm2712-raspberrypi5-domd.dts b/arch/arm64/boot/dts/broadcom/bcm2712-raspberrypi5-domd.dts
+new file mode 100644
+index 000000000000..1457e696f968
+--- /dev/null
++++ b/arch/arm64/boot/dts/broadcom/bcm2712-raspberrypi5-domd.dts
+@@ -0,0 +1,2 @@
++// SPDX-License-Identifier: GPL-2.0
++#include "arm/broadcom/bcm2712-rpi-5-b.dts"
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch]: why do you add chosen to domd device tree? 
Why not use prod-devel approach with xen-chosen.dtsi?

Comment (oleksiimoisieiev) [reply meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch]: using this approach will allow you to make changes to image size on build and do not do this on each boot

File: meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+diff --git a/arch/arm64/boot/dts/broadcom/Makefile b/arch/arm64/boot/dts/broadcom/Makefile
+index 6dd8659a7d8a..a5bf8623194f 100644
+--- a/arch/arm64/boot/dts/broadcom/Makefile
++++ b/arch/arm64/boot/dts/broadcom/Makefile
+@@ -21,6 +21,8 @@ dtb-$(CONFIG_ARCH_BCM2835) += bcm2710-rpi-cm3.dtb
+ dtb-$(CONFIG_ARCH_BCM2835) += bcm2711-rpi-cm4.dtb
+ dtb-$(CONFIG_ARCH_BCM2835) += bcm2711-rpi-cm4s.dtb
+ dtb-$(CONFIG_ARCH_BCM2835) += bcm2712-rpi-5-b.dtb
++dtb-$(CONFIG_ARCH_BCM2835) += bcm2712-raspberrypi5-xen.dtb
```
Comment (oleksiimoisieiev) [inline meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch]: i don't see raspberrypypi5-xen in this patch... Is it created on build?

**Review by oleksiimoisieiev — CHANGES_REQUESTED** [review]

File: meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch (hunk 14 above)
Comment (dsemenets) [reply meta-rpi-domd/recipes-kernel/linux/files/0001-aarch64-dts-Add-device-tree-for-xen.patch]: I will delete this. Rpi bootloader wants only bcm2712-rpi-5-b.dtb

File: meta-rpi-domd/conf/layer.conf (hunk 12 above)
Comment (dsemenets) [reply meta-rpi-domd/conf/layer.conf]: Deleted

**Comment by dsemenets** [comment]
Add load xenpolicy

**Pull request closed without merging**